services/app-api/src/postgres/contractAndRates/updateDraftContractRates.ts

```diff
@@ -0,0 +1,259 @@
+                    position: idx,
+                    ...d,
+                })),
+        },
+        certifyingActuaryContacts: {
+            deleteMany: {},
+            create:
+                rateFormData.certifyingActuaryContacts &&
+                rateFormData.certifyingActuaryContacts.map((c, idx) => ({
+                    position: idx,
+                    ...c,
+                })),
+        },
+        addtlActuaryContacts: {
+            deleteMany: {},
+            create:
+                rateFormData.addtlActuaryContacts &&
+                rateFormData.addtlActuaryContacts.map((c, idx) => ({
+                    position: idx,
+                    ...c,
+                })),
+        },
+        actuaryCommunicationPreference:
+            rateFormData.actuaryCommunicationPreference,
+    }
+}
+
+async function updateDraftContractRates(
+    client: PrismaTransactionType,
+    args: UpdateDraftContractRatesArgsType
+): Promise<ContractType | Error> {
+    try {
+        // for now, get the latest contract revision, eventually we'll have rate revisions directly on this
+        const contract = await client.contractTable.findUnique({
+            where: {
+                id: args.contractID,
+            },
+            include: {
+                revisions: {
+                    take: 1,
+                    orderBy: {
+                        createdAt: 'desc',
+                    },
+                },
+            },
+        })
+
+        if (!contract) {
+            return new NotFoundError(
+                'contract not found with ID: ' + args.contractID
+            )
+        }
+
+        const draftRevision = contract.revisions[0]
+        if (!draftRevision) {
+            return new Error(
+                'PROGRAMMER ERROR: This draft contract has no draft revision'
+            )
+        }
+
+        // figure out the rate number range for created rates.
+        const state = await client.state.findUnique({
+            where: { stateCode: contract.stateCode },
+        })
+
+        if (!state) {
+            return new Error(
+                'PROGRAMER ERROR: No state found with code: ' +
+                    contract.stateCode
+            )
+        }
+
+        let nextRateNumber = state.latestStateRateCertNumber
+
+        // create new rates with new revisions
+        const ratesToCreate = args.rateUpdates.create.map((ru) => {
+            const rateFormDataem = ru.formData
+            const thisRateNumber = nextRateNumber
+            nextRateNumber++
+            return {
+                stateCode: contract.stateCode,
+                stateNumber: thisRateNumber,
+                revisions: {
+                    create: prismaRateCreateFormDataFromDomain(rateFormDataem),
+                },
+            }
+        })
+
+        // to delete draft rates, we need to delete their revisions first
+        await client.rateRevisionTable.deleteMany({
```

Comment:
Do we want to wrap this `deleteMany` and the `update` in a transaction? If an `update` on the contract revision fails we probably want to roll back the draft rate deletions?